<a href="https://colab.research.google.com/github/ValentinoVizner/Python_Deep_Dive_1/blob/master/Section_4_Numeric_Types.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NUMBERS

## 1.Introduction

![alt text](https://drive.google.com/uc?id=16x9jU5dUcuW5nk7p-pfzbLkSbFwgRif0)
</br>
So in Python we have 5 main types of numbers, while in world of Mathematics we have 4 main types of numbers.

## 2.Integers: Data Types

In [0]:
!pip install ipython-autotime
%load_ext autotime

  Created wheel for ipython-autotime: filename=ipython_autotime-0.1-cp36-none-any.whl size=1832 sha256=c9a4aebeeccf376b9b80630178141c25ac044c82009c838482c59f33ce357dd6
  Stored in directory: /root/.cache/pip/wheels/d2/df/81/2db1e54bc91002cec40334629bc39cfa86dff540b304ebcd6e
Successfully built ipython-autotime


![alt text](https://drive.google.com/uc?id=1IVVWdHvThhWDsIeoc4Vo4Ib-_6UQlVb1)

![alt text](https://drive.google.com/uc?id=1qJC10_kVw0FLcexNAQic1unIdNi15YLS)

![alt text](https://drive.google.com/uc?id=1RsNkNcV79VvCL4ehfZdnCrPNSgsJcr55)

![alt text](https://drive.google.com/uc?id=13wvnLZHc9ExBmWTrze23Pt2qSTKNm5ow)

![alt text](https://drive.google.com/uc?id=1sS7CqYRcGP-iAAh1cQiv-I_r3CcrzfBV)
Be cautious when dealing with a lot of operations, try to see biggest numbers and limit them to certain length.

Integers are objects - instances of the ``int`` class.

In [0]:
print(type(100))

<class 'int'>
time: 2.83 ms


They are a variable length data type that can theoretically handle any integer magnitude. This will take up a variable amount of memory that depends on the particular size of the integer.

In [0]:
import sys

sys.getsizeof(0)

24

In [0]:
sys.getsizeof(2**1000
            )

160

time: 8.17 ms


In [0]:
(160 
 - 24 # number of overhead, bytes
 ) * 8

1088

time: 3.19 ms


In [0]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

time: 3.42 ms


Larger integers will also slow down calculations.

In [0]:
import time

def calc(a):
    for i in range(10000000):
        a * 2
# We start with a small integer value for a (10):

time: 5.1 ms


In [0]:
start = time.perf_counter()
calc(10)
end = time.perf_counter()
print(end - start)

0.4009422130002349
time: 403 ms


In [0]:
start = time.perf_counter()
calc(2**100)
end = time.perf_counter()
print(end - start)

0.6867921570001272
time: 689 ms


In [0]:
start = time.perf_counter()
calc(2**10000)
end = time.perf_counter()
print(end - start)

4.261457938000149
time: 4.26 s


## 3.Integers: Operations

![alt text](https://drive.google.com/uc?id=1Sxws5N8OoR0OjGVp13WSP8Kz5TBj-htK)

![alt text](https://drive.google.com/uc?id=1sW-4cBV3zl3da16mQ2FIBz-lptFH1130)

![alt text](https://drive.google.com/uc?id=15PoAU8sEnaqZALbBg0esrypvUPx1K6SJ)

![alt text](https://drive.google.com/uc?id=1Go-u-Kdd9WSTsXiYQplzgbfKvHkaOmfb)
![alt text](https://drive.google.com/uc?id=1cshyv_Z4asG7FpNJOL-wHH0KVUiWyQAq)

![alt text](https://drive.google.com/uc?id=1rBbU56g585A7UJDQJjMkcxrqZxC9rGL6)

In [0]:
# standard division operator `/` **always** results in a float value.
type(10 / 2)

float

In [0]:
import math
math.floor(3.14)

3

In [0]:
math.floor(-3.60)

-4

In [0]:
math.floor(-3.0005)

-4

In [0]:
math.floor(-3.0000000000000001)

-3

Wait what is ging on!?
</br>
It is because of limited precision for float in Python

#### 3.1. The Floor Division Operator

In [0]:
a = 33
b = 16
print(a/b)
print(a//b)
print(math.floor(a/b))

2.0625
2
2


For positive numbers, `a//b` is basically the same as truncating (taking the integer portion) of `a / b`.
</br>
But this is **not** the case for negative numbers.

In [0]:
a = -33
b = 16
print('{0}/{1} = {2}'.format(a, b, a/b))
print('trunc({0}/{1}) = {2}'.format(a, b, math.trunc(a/b)))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('floor({0}//{1}) = {2}'.format(a, b, math.floor(a/b)))

-33/16 = -2.0625
trunc(-33/16) = -2
-33//16 = -3
floor(-33//16) = -3


In [0]:
a = 33
b = -16
print('{0}/{1} = {2}'.format(a, b, a/b))
print('trunc({0}/{1}) = {2}'.format(a, b, math.trunc(a/b)))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('floor({0}//{1}) = {2}'.format(a, b, math.floor(a/b)))

33/-16 = -2.0625
trunc(33/-16) = -2
33//-16 = -3
floor(33//-16) = -3


#### 3.2. The Modulo Operator

The modulo operator and the floor division operator will always satisfy the following equation:

``a = b * (a // b) + a % b``

In [0]:
a = 13
b = 4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

13/4 = 3.25
13//4 = 3
13%4 = 1
True


In [0]:
a = -13
b = 4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

-13/4 = -3.25
-13//4 = -4
-13%4 = 3
True


In [0]:
a = 13
b = -4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

13/-4 = -3.25
13//-4 = -4
13%-4 = -3
True


In [0]:
a = -13
b = -4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

-13/-4 = 3.25
-13//-4 = 3
-13%-4 = -1
True


## 4.INTEGERS: Constructor and Bases (Part 1)

![alt text](https://drive.google.com/uc?id=1ACGOeXMVdvHnCqlvZQgYDbWEY3qQp1vC)

![alt text](https://drive.google.com/uc?id=1A6VDdgyDk-bHmYiwRNFqq7caeqrLzYbr)

It is BETTER to use explicitly `base=16` for example.
</br>
Max. number of base is **36** because there are 26 letters **[a-z]** and 10 numbers **[0-9]**

![alt text](https://drive.google.com/uc?id=1TaXefPpnz_U3TqcJ0FcyenVu25jSIKGH)

![alt text](https://drive.google.com/uc?id=1MdkyubAKViBIuU9KfBrKZvGOq744kQAM)

![alt text](https://drive.google.com/uc?id=1an_d9qDgVQwsz-RZaHR2UnER9u2buTxV)

![alt text](https://drive.google.com/uc?id=1XrhnusfQq0GWTT6jmqu21NZVDaH-8oST)


![alt text](https://drive.google.com/uc?id=19NeBUnctfHItoPm9WEJRiXhBPT1uHLMz)

![alt text](https://drive.google.com/uc?id=1aDEbdAo_zbU7HHzsQ9d8BpYSvx38ptGj)

### 4.INTEGERS: Constructor and Bases (Part 2) => Exercises

In [0]:
from fractions import Fraction

a = Fraction(22 / 7)

print(float(a))
print(int(a))

3.142857142857143
3
time: 2.89 ms


In [0]:
int("101", base=2)

5

time: 7.95 ms


Python uses ``a-z`` for bases from 11 to 36.
</br>
Note that the letters are not case sensitive.

In [0]:
print(int("F1A", base=16))
print(int("f1a", base=16))

3866
3866
time: 1.94 ms


#### Base Representations

In [0]:
print(f'binary representation of 10: {bin(10)}')
print(f'octaal representation of 10: {oct(10)}')
print(f'hexadcimal representation of 10: {hex(10)}')

binary representation of 10: 0b1010
octaal representation of 10: 0o12
hexadcimal representation of 10: 0xa
time: 3.11 ms


Note the `0b`, `0o` and `0x` prefixes

In [0]:
a = int('1010', 2)
b = int('0b1010', 2)
c = 0b1010
print(a, b, c)

10 10 10
time: 3.15 ms


For literals, the ``a-z`` characters are not case-sensitive either

In [0]:
a = 0xf1a
b = 0xF1a
c = 0xF1A
print(a, b, c)

3866 3866 3866
time: 6.45 ms


#### Custom Rebasing

Python only provides built-in function to rebase to base 2, 8 and 16.

For other bases, you have to provide your own algorithm (or leverage some 3rd party library of your choice)

In [0]:
def from_base10(n, b):
    if b < 2:
        raise ValueError('Base b must be >= 2')
    if n < 0:
        raise ValueError('Number n must be >= 0')
    if n == 0:
        return [0]
    digits = []
    while n > 0:
        # m = n % b
        # n = n // b
        # which is the same as:
        n, m = divmod(n, b)
        digits.insert(0, m)
    return digits

time: 9.17 ms


In [0]:
from_base10(255, 16)

[15, 15]

time: 2.9 ms


Next we may want to encode the digits into strings using different characters for each digit in the base

In [0]:
def encode(digits, digit_map):
    # we require that digit_map has at least as many
    # characters as the max number in digits
    if max(digits) >= len(digit_map):
        raise ValueError("digit_map is not long enough to encode digits")
    
    # we'll see this later, but the following would be better:
    return ''.join([digit_map[d] for d in digits])
    
    

time: 2.33 ms


In [0]:
encode([1, 0, 1], "FT")

'TFT'

time: 3.96 ms


In [0]:
encode([1, 10, 11], '0123456789AB')

'1AB'

time: 3.21 ms


And we can combine both functions into a single one for easier use:

In [0]:
def rebase_from10(number, base):
    digit_map = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    if base < 2 or base > 36:
        raise ValueError('Invalid base: 2 <= base <= 36')
    # we store the sign of number and make it positive
    # we'll re-insert the sign at the end
    sign = -1 if number < 0 else 1
    number *= sign
    
    digits = from_base10(number, base)
    encoding = encode(digits, digit_map)
    if sign == -1:
        encoding = '-' + encoding
    return encoding

time: 6.47 ms


In [0]:
e = rebase_from10(10, 2)
print(e)
print(int(e, 2))

1010
10
time: 2.5 ms


In [0]:
e = rebase_from10(-10, 2)
print(e)
print(int(e, 2))

-1010
-10
time: 2.7 ms


In [0]:
rebase_from10(131, 11)

'10A'

time: 3.42 ms


In [0]:
rebase_from10(4095, 16)

'FFF'

time: 2.33 ms


In [0]:
rebase_from10(-4095, 16)

'-FFF'

time: 3.12 ms


## 5.Rational Numbers

![alt text](https://drive.google.com/uc?id=1sH3CSU3YVO-4K2wIdgJCD5B5DHE6046V)

![alt text](https://drive.google.com/uc?id=1nwUA3tPA8qkqmtFJ4WUYevIXrRef95_w)

![alt text](https://drive.google.com/uc?id=1N5DJi0XzPxpkudrMCpzpfFCPEgtbnIjD)

![alt text](https://drive.google.com/uc?id=11P5Ir0N8jk9dQp0WriTltyk2cSSH239v)

![alt text](https://drive.google.com/uc?id=1EB_7hUEZwR_XaEMhDIrbnuKV5S4NH_xC)

![alt text](https://drive.google.com/uc?id=1iYbg5uc-DpJ_lOoBPAeROV4RIt3ng_Bz)

![alt text](https://drive.google.com/uc?id=1SmVsT0QnFPeGPkygrByKw9PN7uW9Kxf3)

### Coding Time

In [0]:
from fractions import Fraction
help(Fraction)

Help on class Fraction in module fractions:

class Fraction(numbers.Rational)
 |  This class implements rational numbers.
 |  
 |  In the two-argument form of the constructor, Fraction(8, 6) will
 |  produce a rational number equivalent to 4/3. Both arguments must
 |  be Rational. The numerator defaults to 0 and the denominator
 |  defaults to 1 so that Fraction(3) == 3 and Fraction() == 0.
 |  
 |  Fractions can also be constructed from:
 |  
 |    - numeric strings similar to those accepted by the
 |      float constructor (for example, '-2.3' or '1e10')
 |  
 |    - strings of the form '123/456'
 |  
 |    - float and Decimal instances
 |  
 |    - other Rational instances (including integers)
 |  
 |  Method resolution order:
 |      Fraction
 |      numbers.Rational
 |      numbers.Real
 |      numbers.Complex
 |      numbers.Number
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __abs__(a)
 |      abs(a)
 |  
 |  __add__(a, b)
 |      a + b
 |  
 |  __bool__(a)
 |

We can create Fraction objects in a variety of ways:

In [0]:
print(Fraction(1))
print(Fraction(numerator=1, denominator=3))
print(Fraction(denominator=3, numerator=1))
x = Fraction(2, 3)
y = Fraction(3, 4)
# 2/3 / 3/4 --> 2/3 * 4/3 --> 8/9
print(Fraction(x, y))

1
1/3
1/3
8/9


In [0]:
Fraction('22/7')

Fraction(22, 7)

In [0]:
Fraction('10')

Fraction(10, 1)

Fractions are automatically reduced:

In [0]:
Fraction(8, 16)

Fraction(1, 2)

Negative sign is attached to the numerator:

In [0]:
x = Fraction(1, -4)
x

Fraction(-1, 4)

In [0]:
print(x.numerator, x.denominator)

-1 4


Standard arithmetic operators are supported:

In [0]:
Fraction(1, 3) + Fraction(1, 3) + Fraction(1, 3)

Fraction(1, 1)

In [0]:
Fraction(1, 2) * Fraction(1, 4)

Fraction(1, 8)

In [0]:
Fraction(1, 2) / Fraction(1, 3)

Fraction(3, 2)

In [0]:
import math

x = Fraction(math.pi)
x

Fraction(884279719003555, 281474976710656)

In [0]:
y = Fraction(math.sqrt(2))
y

Fraction(6369051672525773, 4503599627370496)

In [0]:
float(y)

1.4142135623730951

Note that these rational values are approximations to the irrational numbers $\pi$ and $\sqrt{2}$

**Beware!!**
</br>
Float number representations (as we will examine in future lessons) do not always have an exact representation.
</br>
The number 0.125 (1/8) **has** an exact representation:

In [0]:
a = 0.125
b = 0.3
print(a)
Fraction(a)

0.125


Fraction(1, 8)

In [0]:
print(b)
Fraction(b)

0.3


Fraction(5404319552844595, 18014398509481984)

and so we see the expected equivalent fraction.

But, 0.3 (3/10) does **not** have an exact representation.
</br>
We will study this in upcoming lessons.

But for now, let's just see a quick explanation:

In [0]:
x = 0.3
print(x)

0.3


Everything looks ok here - why am I saying 0.3 (float) is just an approximation?

Python is trying to format the displayed value for readability - so it rounds the number for a better display format!

We can instead choose to display the value using a certain number of digits:

In [0]:
format(b, '5f')

'0.300000'

At 5 digits after the decimal, we might still think 0.3 is an exact representation.

But let's display a few more digits:

In [0]:
format(x, '.25f')

'0.2999999999999999888977698'

Now we see that **x** is not quite 0.3...
</br>
</br>
In fact, we can quantify the delta this way:

In [0]:
delta = Fraction(0.3) - Fraction(3, 10)

Theoretically, delta should be 0, but it's not:


In [0]:
delta == 0

False

**delta** is a very small number, the above fraction...

As a float:

In [0]:
float(delta)

-1.1102230246251566e-17

In [0]:
x = Fraction(math.pi)
print(x)
print(format(float(x), '.25f'))

884279719003555/281474976710656
3.1415926535897931159979635


In [0]:
y = x.limit_denominator(10)
print(y)
print(format(float(y), '.25f'))

22/7
3.1428571428571427937015414


In [0]:
y = x.limit_denominator(10000)
print(y)
print(format(float(y), '.25f'))

355/113
3.1415929203539825209645642


## 6.FLOATS: Internal Representation (Theory)

![alt text](https://drive.google.com/uc?id=1dgxa5uzAoAIqGesx-Ct_k80IX0IDESUP)

All these significant digits have only 5 significant digist.
</br>
It might seem that number `12345000000` has more then 5, but it has actually 5 significant digits 12345 and times 10e6. (12345 * 10e6) 

![alt text](https://drive.google.com/uc?id=16U-XQnBR3WXxt_x6T0VOjCBkkBoUDX2X)

![alt text](https://drive.google.com/uc?id=1mGXNjI1o3X9SNWIVN7KizbH7WJBsFrfL)

![alt text](https://drive.google.com/uc?id=1lzvWwCMe2bF3JhViYSZ6EGyVtDsjIIgA)

![alt text](https://drive.google.com/uc?id=1PLeQveRWtYgwIZ_D7kqlZkd1NrmuaHJe)

![alt text](https://drive.google.com/uc?id=1V8VUAIWA96hEbDrp1eqQn8ngwRkdtbr5)

### Coding Time 

The ``float`` class has a single constructor, which can take a number or a string and will attempt to convert it to a float.

In [0]:
print(float(3.14))
print(float('10'))
print(float('0.1'))

3.14
10.0
0.1


In [0]:
float('22/7')

ValueError: ignored

If you really want to get a float from a string such as ``'22/7'``, you could first create a ``Fraction``, then create a ``float`` from that:

In [0]:
from fractions import Fraction
float(Fraction('22/7'))

3.142857142857143

Floats do not always have an exact representation:

In [0]:
print(0.1)

0.1


### INTERESTING FACT: Although this looks like ``0.1`` exactly, we need to reveal more digits after the decimal point to see what's going on:

In [0]:
format(0.1, '.25f')

'0.1000000000000000055511151'

However, certain numbers can be represented exactly in a binary fraction expansion:

In [0]:
format(0.125, '.25f')

'0.1250000000000000000000000'

This is because 0.125 is precisely 1/8, or 1/(2^3)
</br>
</br>
</br>
This is interesting though

In [0]:
a = 0.1 + 0.1 + 0.1
b = 0.3

In [0]:
a == b

False

### INTERESTING FACT: WAIT **WHAAAAAAAAAAAT** what is going on???!?!?!?!?!

In [0]:
format(a, '.25f')

'0.3000000000000000444089210'

In [0]:
format(b, '.25f')

'0.2999999999999999888977698'

SO here we can see what is going on. Since computers use binary, we see that `0.1` in binary, i.e. float precision is not accurate like in decimal so we can not get `0.1` in binary base.

## 7.FLOATS: Equality Testing

![alt text](https://drive.google.com/uc?id=1NyZ5b4I-XBkAqu1ErxlP39yD4Kb1j3-q)

![alt text](https://drive.google.com/uc?id=1P1hA5mNrFndGRyoJKEMVAOvLP6AOby5B)

![alt text](https://drive.google.com/uc?id=1zHqab1WMsXXPL375_YDj9c5sOYuRZ9Jl)

![alt text](https://drive.google.com/uc?id=1FL0VXUmaXYVJgOUiAeoQQ65xH4MDv-96)

![alt text](https://drive.google.com/uc?id=1eMJNHQD4RWyeEwpa5ZycuWlD0bsRp8jv)

![alt text](https://drive.google.com/uc?id=1ioXOzhLyqAqqjweufWVJNqCxUee2BouC)

![alt text](https://drive.google.com/uc?id=1rLW4_VwtNYpT84nIlbk9RgFNWST9xTkG)
When you look at first example, `x` and `y`, `0.01` isn't actually a big difference and they are close to each other.
</br>
However when you look and `a` and `b` they are relativly speaking not so close, it is a big difference when we compare the 2, so it is good that we get a `True` from the `math.isclose(...)`

### Coding Time

Because not all real numbers have an exact ``float`` representation, equality testing can be tricky.

In [0]:
x = 0.1 + 0.1 + 0.1
y = 0.3
x == y

False

This is because ``0.1`` and ``0.3`` do not have exact representations:

In [0]:
print('0.1 --> {0:.25f}'.format(0.1))
print('x --> {0:.25f}'.format(x))
print('y --> {0:.25f}'.format(y))

0.1 --> 0.1000000000000000055511151
x --> 0.3000000000000000444089210
y --> 0.2999999999999999888977698


However, in some (limited) cases where all the numbers involved do have an exact representation, it will work:

In [0]:
x = 0.125 + 0.125 + 0.125
y = 0.375
x == y

True

In [0]:
print('0.125 --> {0:.25f}'.format(0.125))
print('x --> {0:.25f}'.format(x))
print('y --> {0:.25f}'.format(y))

0.125 --> 0.1250000000000000000000000
x --> 0.3750000000000000000000000
y --> 0.3750000000000000000000000


One simple way to get around this is to round to a specific number of digits and then compare

In [0]:
x = 0.1 + 0.1 + 0.1
y = 0.3
round(x, 5) == round(y, 5)

True

We can also use a more flexible technique implemented by the ``isclose`` method in the ``math`` module

In [0]:
from math import isclose

help(isclose)

Help on built-in function isclose in module math:

isclose(...)
    isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0) -> bool
    
    Determine whether two floating point numbers are close in value.
    
       rel_tol
           maximum difference for being considered "close", relative to the
           magnitude of the input values
        abs_tol
           maximum difference for being considered "close", regardless of the
           magnitude of the input values
    
    Return True if a is close in value to b, and False otherwise.
    
    For the values to be considered close, the difference between them
    must be smaller than at least one of the tolerances.
    
    -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
    is, NaN is not close to anything, even itself.  inf and -inf are
    only close to themselves.



In [0]:
x = 0.1 + 0.1 + 0.1
y = 0.3
isclose(x, y)

True

The ``isclose`` method takes two optional parameters, ``rel_tol`` and ``abs_tol``.

``rel_tol`` is a relative tolerance that will be relative to the magnitude of the largest of the two numbers being compared. Useful when we want to see if two numbers are close to each other as a percentage of their magnitudes.

``abs_tol`` is an absolute tolerance that is independent of the magnitude of the numbers we are comparing - this is useful for numbers that are close to zero.

In this situation we might consider x and y to be close to each other:

In [0]:
# Here they are close to each other
x = 123456789.01
y = 123456789.02
y / x

1.0000000000809999


but not in this case:

In [0]:
# They are not close to each other, actually y i 2 times bigger
x = 0.01
y = 0.02

y / x

2.0


 When we use round we can see how can `round()` be dangerous

In [0]:
round(x, 1) == round(y, 1)

True

In both these cases the difference between the two numbers was ``0.01``, yet in one case we considered the numbers "equal" and in the other, not "equal". Relative tolerances are useful to handle these scenarios.

In [0]:
isclose(123456789.01, 123456789.02, rel_tol=0.01)

True

In [0]:
isclose(0.01, 0.02, rel_tol=0.01)

False

### INTERESTING FACT: On the other hand, we have to be careful with relative tolerances when working with values that are close to zero:

In [0]:
x = 0.0000001
y = 0.0000002
isclose(x, y, rel_tol=0.01)

False

So, we could use an absolute tolerance here:

In [0]:
isclose(x, y, abs_tol=0.0001, rel_tol=0)

True

In general, we can combine the use of both relative and absolute tolerances in this way:

In [0]:
x = 0.0000001
y = 0.0000002

a = 123456789.01
b = 123456789.02

print('x = y:', isclose(x, y, abs_tol=0.0001, rel_tol=0.01))
print('a = b:', isclose(a, b, abs_tol=0.0001, rel_tol=0.01))

x = y: True
a = b: True


# 8.FLOATS: Coercing to Integers - Lecture

![alt text](https://drive.google.com/uc?id=14aRwcow6Y1Q5BIk40pYU9Jf9pTgiKyGK)
![alt text](https://drive.google.com/uc?id=1XChg7kQmSsDi39gd3Esd9k0582w9fPTj)

![alt text](https://drive.google.com/uc?id=1g1-wUr4lWCaIiYdKolu6CcJJhfwbfHxT)

![alt text](https://drive.google.com/uc?id=13Up9O5loJfxNC6WbTPUnjyOJMqx75nXO)

![alt text](https://drive.google.com/uc?id=1Ag9TTXDTpqEZcQ-4pkdwBZwB5vOp69J2)

![alt text](https://drive.google.com/uc?id=1piVggY9gyuNNz_U7WjSRYtGhmI07BrZK)

## Coding Time

### Truncation

In [0]:
from math import trunc
trunc(10.3), trunc(10.5), trunc(10.6)    

(10, 10, 10)

In [0]:
trunc(-10.6), trunc(-10.5), trunc(-10.3)

(-10, -10, -10)

The **int** constructor uses truncation when a float is passed in:

In [0]:
int(10.3), int(10.5), int(10.6)

(10, 10, 10)

In [0]:
int(-10.5), int(-10.5), int(-10.4)

(-10, -10, -10)

### Floor

In [0]:
from math import floor
floor(10.4), floor(10.5), floor(10.6)

(10, 10, 10)

In [0]:
floor(-10.4), floor(-10.5), floor(-10.6)

(-11, -11, -11)

### Ceiling

In [0]:
from math import ceil
ceil(10.4), ceil(10.5), ceil(10.6)

(11, 11, 11)

In [0]:
ceil(-10.4), ceil(-10.5), ceil(-10.6)

(-10, -10, -10)

# 9.FLOAT: Rounding

![alt text](https://drive.google.com/uc?id=13c0Nszf3xigBxx3eTcim5czyCsaR7pps)

![alt text](https://drive.google.com/uc?id=1FietmzsWr8e88FW6mzcQQ_jUnJvyJAwX)

![alt text](https://drive.google.com/uc?id=102ctyHxncWE_lwAc1jwRgHEaaYqVexr0)

![alt text](https://drive.google.com/uc?id=1ucbOX04o3Td0rwrTivuIXTmKUqlsU0Re)

![alt text](https://drive.google.com/uc?id=1B7ST4pYjxyalRJEBsc8bJ9pXYwjKq8HB)

![alt text](https://drive.google.com/uc?id=1M61rNUjUIiLk9fXn_JMOMNNlWIvoAdqZ)

![alt text](https://drive.google.com/uc?id=1RbPm2w8nCilFNLGwy51U6sKwansC5m3j)

![alt text](https://drive.google.com/uc?id=1My2aBvY6c6wbZ1XjdlVOCCCW-B6UncqT)

![alt text](https://drive.google.com/uc?id=1mei77rdrMzNmy0E8dPnzdBTXS8z55jlh)

![alt text](https://drive.google.com/uc?id=1qCbjArRCg65nXcSkXcVAgnO2blWj17OR)

![alt text](https://drive.google.com/uc?id=1SbMyUv35byBRRdx0c-cvmRINLFgChqTC)

## Coding Time

### Rounding

In [0]:
help(round)

Help on built-in function round in module builtins:

round(...)
    round(number[, ndigits]) -> number
    
    Round a number to a given precision in decimal digits (default 0 digits).
    This returns an int when called with one argument, otherwise the
    same type as the number. ndigits may be negative.



#### n = 0

In [0]:
a = round(1.5)
a, type(a)

(2, int)

In [0]:
a = round(1.5, 0)
a,  type(a)

(2.0, float)

#### n > 0

In [0]:
round(1.88888, 3), round(1.8888, 2), round(1.8888, 1), round(1.8888, 0)

(1.889, 1.89, 1.9, 2.0)

#### n < 0

In [0]:
round(888.88, 1), round(888.88, 0), \
round(888.88, -1), round(888.88, -2), round(888.88, -3), round(888.88, -4)

(888.9, 889.0, 890.0, 900.0, 1000.0, 0.0)

#### Ties

In [0]:
round(1.25, 1)

1.2

In [0]:
round(1.35, 1)

1.4

This is rounding to nearest, with ties to nearest number with even least significant digit, aka Banker's Rounding.
</br>
</br>
Works similarly with **n** negative.

In [0]:
print(round(15, -1))
print(round(25, -1))

20
20


#### Rounding to closest, ties away from zero

This is traditionally the type of rounding taught in school, which is different from the Banker's Rounding implemented in Python (and in many other programming languages)
</br> 
</br>
1.5 --> 2 <br>
2.5 --> 3 <br>

-1.5 --> -2 <br>
-2.5 --> -3 <br>
</br>
</br>
To do this type of rounding (to nearest 1) we can add (for positive numbers) or subtract (for negative numbers) 0.5 and then truncate the resulting number.

In [0]:
def _round(x):
    from math import copysign
    return int(x + 0.5 * copysign(1, x))

In [0]:
print(round(1.5), _round(1.5))
print(round(2.5), _round(2.5))

2 2
2 3


# 10.DECIMALS

![alt text](https://drive.google.com/uc?id=1IRwhcS0jtnnkgU8xP7bka9ushUuBOuz6)

![alt text](https://drive.google.com/uc?id=1JhIeJ9926gkB-FO5VT3TAcgl0ZL1P8EJ)

![alt text](https://drive.google.com/uc?id=1TLzo1I_gY1U-Sr-jnIInrPZE50wTJEYP)

![alt text](https://drive.google.com/uc?id=1O_7P5BPuopyXdZjprfylUB-JtJb98ozx)

![alt text](https://drive.google.com/uc?id=1gN49UTbhfsmbn0WZOXLRES7Sw5-orx1p)

## Coding Time

In [0]:
import decimal
from decimal import Decimal

Decimals have context, that can be used to specify rounding and precision (amongst other things)
</br>
</br>
Contexts can be local (temporary contexts) or global (default)

### Global Context

In [0]:
g_ctx  = decimal.getcontext()
g_ctx.prec

28

In [0]:
print(type(g_ctx))
print(type(decimal.localcontext()))


<class 'decimal.Context'>
<class 'decimal.ContextManager'>


This HINTS we have to use `with` statement

In [0]:
g_ctx.rounding

'ROUND_HALF_EVEN'

In [0]:
g_ctx.prec = 6
g_ctx.rounding = decimal.ROUND_HALF_UP


In [0]:
decimal.getcontext()

Context(prec=6, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[Inexact, Rounded], traps=[InvalidOperation, DivisionByZero, Overflow])

And if we read this back directly from the global context:

In [0]:
decimal.getcontext().prec

6

In [0]:
decimal.getcontext().rounding

'ROUND_HALF_UP'

### Local Context

The ``localcontext()`` function will return a context manager that we can use with a ``with`` statement:

In [0]:
with decimal.localcontext() as ctx:
    print(ctx.prec)
    print(ctx.rounding)
    print(id(ctx) == id(decimal.getcontext()))

6
ROUND_HALF_EVEN
True


Since no argument was specified in the ``localcontext()`` call, it provides us a context manager that uses a copy of the global context.
</br>
</br>
Modifying the local context has no effect on the global context

In [0]:
with decimal.localcontext() as ctx:
    ctx.prec = 10
    print('local prec = {0}, global prec = {1}'.format(ctx.prec, g_ctx.prec))

local prec = 10, global prec = 6


### Rounding

In [0]:
decimal.getcontext().rounding

'ROUND_HALF_UP'

The rounding mechanism is ROUND_HALF_UP because we set the global context to that earlier in this notebook. Note that normally the default is ROUND_HALF_EVEN.
</br>
So we first reset our global context rounding to that:

In [0]:
decimal.getcontext().rounding = decimal.ROUND_HALF_EVEN

In [0]:
x = Decimal('1.25')
y = Decimal('1.35')
print(round(x, 1))
print(round(y, 1))

1.2
1.4


Let's change the rounding mechanism in the global context to ROUND_HALF_UP:

In [0]:
decimal.getcontext().rounding = decimal.ROUND_HALF_UP

In [0]:
x = Decimal('1.25')
y = Decimal('1.35')
print(round(x, 1))
print(round(y, 1))

1.3
1.4


As you may have realized, changing the global context is a pain if you need to constantly switch between different precisions and rounding algorithms. Also, it could introduce bugs if you forget that you changed the global context somewhere further up in your module.

For this reason, it is usually better to use a local context manager instead:

First we reset our global context rounding to the default:

In [0]:
decimal.getcontext().rounding = decimal.ROUND_HALF_EVEN

In [0]:
x = Decimal('1.25')
y = Decimal('1.35')
print(round(x, 1), round(y, 1))
with decimal.localcontext() as ctx:
    ctx.rounding = decimal.ROUND_HALF_UP
    print(round(x, 1), round(y, 1))
print(round(x, 1), round(y, 1))

1.2 1.4
1.3 1.4
1.2 1.4


# 11.DECIMALS: Constructors and Contexts

![alt text](https://drive.google.com/uc?id=1J2flbXcDER9Cqd1_hlPlyODbrspG4aiW)

![alt text](https://drive.google.com/uc?id=1A0jTFyFGE0TSko2T5lW3SXZKMDVzHBmt)

![alt text](https://drive.google.com/uc?id=1NZkcYkPkhvoJxp4wFL6tuV8zGYKYoMA-)

![alt text](https://drive.google.com/uc?id=17mmWRsuxjjbtzm6UYeQMlYj4637Z42ja)

## Coding Time

The **Decimal** constructor can handle a variety of data types

In [0]:
import decimal
from decimal import Decimal

### Integers

In [0]:
print(Decimal(10))
print(Decimal(-10))

10
-10


### Strings

In [0]:
print(Decimal('0.1'))
print(Decimal('-3.1415'))

0.1
-3.1415


### Tuples

CAREFUL! You need double `(...)` because it is tuple.

In [0]:
print(Decimal ((0, (3,1,4,1,5), -4)))
print(Decimal((1, (1,2,3,4), -3)))
print(Decimal((0, (1,2,3), 3)))

3.1415
-1.234
1.23E+5


### But don't use Floats

In [0]:
print(format(0.1, '.25f'))
print(Decimal(0.1))

0.1000000000000000055511151
0.1000000000000000055511151231257827021181583404541015625


As you can see, since we passed an approximate binary float to the Decimal constructor it did it's best to represent that binary float **exactly**!!
</br>
</br>
So, instead, use strings or tuples in the Decimal constructor.

### Context Precision and the Constructor

The context precision does nto affect the precision used when creating a Decimal object - those are independent of each other.
</br>
</br>
Let's set our global (default) context to a precision of 2

In [0]:
decimal.getcontext().prec = 2

Now we can create decimal numbers of higher precision than that:

In [0]:
a = Decimal('0.12345')
b = Decimal('0.12345')

print(a,
      b)

0.12345 0.12345


But when we add those two numbers up, the context precision will matter:

In [0]:
a+b

Decimal('0.25')

As you can see, we ended up with a sum that was rounded to 2 digits after the decimal point (precision = 2)

### Local and Global Contexts are Independent

In [0]:
decimal.getcontext().prec = 6
decimal.getcontext().rounding

'ROUND_HALF_EVEN'

In [0]:
a = Decimal('0.12345')
b = Decimal('0.12345')
print(a + b)
with decimal.localcontext() as ctx:
    ctx.prec = 2
    c = a + b
    print('c within local context: {0}'.format(c))
print('c within global context: {0}'.format(c))

0.24690
c within local context: 0.25
c within global context: 0.25


Since **c** was created within the local context by adding **a** and **b**, and the local context had a precision of 2, **c** was rounded to 2 digits after the decimal point.

Once the local context is destroyed (after the **with** block), the variable **c** still exists, and its precision is **still** just 2 - it doesn't magically suddenly get the global context's precision of 6.

# 12.DECIMALS: Math Operators

![alt text](https://drive.google.com/uc?id=1peAOctoNZWmWLnw9SFSyzCSlsqxZ9ojW)

![alt text](https://drive.google.com/uc?id=1f-qFLNou55xB3vUqL-1DzGFFe7KrnKqa)

![alt text](https://drive.google.com/uc?id=1vgYNRpBx55qhzAMnGNBADYBDo6dWG9w1)

![alt text](https://drive.google.com/uc?id=1WfC3SPa94yq2Xf7sxWXZ580aEUyrmrtN)

![alt text](https://drive.google.com/uc?id=1XYyKL-ONecaG1QXgPmNcfvXulYCjs4yZ)

## Coding Time

### Div and Mod

The // and % operators (and consequently, the divmod() function) behave differently for integers and Decimals.
</br>
</br>
This is because integer division for Decimals is performed differently, and results in a truncated division, whereas integers use a floored division.

These differences are only when negative numbers are involved. If all numbers involved are positive, then integer and Decimal div and mod operations are equal.
</br>
</br>
But in both cases the // and % operators satisfy the equation:

``n = d * (n // d) + (n % d)``
</br>
</br>

In [0]:
import decimal
from decimal import Decimal

x = 10
y = 3
print(x//y, x%y)
print(divmod(x, y))
print( x == y * (x//y) + x % y)

3 1
(3, 1)
True


In [0]:
a = Decimal('10')
b = Decimal('3')
print(a//b, a%b)
print(divmod(a, b))
print( a == b * (a//b) + a % b)

3 1
(Decimal('3'), Decimal('1'))
True


As we can see, the // and % operators had the same result when both numbers were positive.

In [0]:
x = -10
y = 3
print(x//y, x%y)
print(divmod(x, y))
print( x == y * (x//y) + x % y)

-4 2
(-4, 2)
True


In [0]:
a = Decimal('-10')
b = Decimal('3')
print(a//b, a%b)
print(divmod(a, b))
print( a == b * (a//b) + a % b)

-3 -1
(Decimal('-3'), Decimal('-1'))
True


On the other hand, we see that in this case the // and % operators did not result in the same values, although the equation was satisfied in both instances.

### Other Mathematical Functions

The Decimal class implements a variety of mathematical functions.

In [0]:
a = Decimal('1.5')
print(a.log10())  # base 10 logarithm
print(a.ln())     # natural logarithm (base e)
print(a.exp())    # e**a
print(a.sqrt())   # square root

0.176091
0.405465
4.48169
1.22474


Although you can use the math function of the math module, be aware that the math module functions will cast the Decimal numbers to floats when it performs the various operations. So, if the precision is important (which it probably is if you decided to use Decimal numbers in the first place), choose the math functions of the Decimal class over those of the math module.

In [0]:
x = 2
x_dec = Decimal(2)

In [0]:
import math

root_float = math.sqrt(x) # method from math
root_mixed = math.sqrt(x_dec) 
root_dec = x_dec.sqrt() # method from Decimal

In [0]:
print(format(root_float, '1.27f'))
print(format(root_mixed, '1.27f'))
print(root_dec)

1.414213562373095145474621859
1.414213562373095145474621859
1.41421


In [0]:
print(format(root_float * root_float, '1.27f'))
print(format(root_mixed * root_mixed, '1.27f'))
print(root_dec * root_dec)

2.000000000000000444089209850
2.000000000000000444089209850
1.99999


#### IMPORTANT: We have to pass `string` to Decimal

In [0]:
x = 0.01
x_dec = Decimal('0.01')

root_float = math.sqrt(x)
root_mixed = math.sqrt(x_dec)
root_dec = x_dec.sqrt()

print(format(root_float, '1.27f'))
print(format(root_mixed, '1.27f'))
print(root_dec)

0.100000000000000005551115123
0.100000000000000005551115123
0.1


In [0]:
print(format(root_float * root_float, '1.27f'))
print(format(root_mixed * root_mixed, '1.27f'))
print(root_dec * root_dec)

0.010000000000000001942890293
0.010000000000000001942890293
0.01


# 13.DECIMALS: Performance Considerations

![alt text](https://drive.google.com/uc?id=16n_6OWUU7DPhzA62lARGWwiu_rydTet6)

## Coding Time



### Memory Footprint

Decimals take up a lot more memory than floats.

In [0]:
import sys
from decimal import Decimal

a = 3.1415
b = Decimal('3.1415')

In [0]:
sys.getsizeof(a)

24

24 bytes are used to store the float 3.1415

In [0]:
sys.getsizeof(b)

104

104 bytes are used to store the Decimal 3.1415

### Computational **Performance**

Decimal arithmetic is also much slower than float arithmetic (on a CPU, and even more so if using a GPU)
</br>
</br>
We can do some rough timings to illustrate this.
</br>
</br>
First we look at the performance difference creating floats vs decimals:

In [0]:
import time
from decimal import Decimal

def run_float(n=1):
    for i in range(n):
        a = 3.1415
        
def run_decimal(n=1):
    for i in range(n):
        a = Decimal('3.1415')


Timing float and Decimal operations:

In [0]:
n = 10000000

start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('decimal: ', end-start)

float:  0.26107951599988155
decimal:  2.4094170359994678


We make a slight variant here to see how addition compares between the two types:

In [0]:
def run_float(n=1):
    a = 3.1415
    for i in range(n):
        a + a
        
def run_decimal(n=1):
    a = Decimal('3.1415')
    for i in range(n):
        a + a
        
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('decimal: ', end-start)

float:  0.4037049119997391
decimal:  0.9386818150005638


How about square roots:

(We drop the n count a bit)

In [0]:
n = 5000000

import math

def run_float(n=1):
    a = 3.1415
    for i in range(n):
        math.sqrt(a)
        
def run_decimal(n=1):
    a = Decimal('3.1415')
    for i in range(n):
        a.sqrt()
        
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('decimal: ', end-start)

float:  0.44518902500021795
decimal:  4.972107605999554


# 14.COMPLEX numbers

![alt text](https://drive.google.com/uc?id=1DI_P7Yn7_5iUmSVDEwh6xKCDDoqygzzW)

![alt text](https://drive.google.com/uc?id=1qPROZOXwXuw4qFfp-Drr8DPmWtrdW7rp)

![alt text](https://drive.google.com/uc?id=1CpLHgVcFTV6S2OVWdHFNJuzjEN1UkqaT)

![alt text](https://drive.google.com/uc?id=1BB0m35lWrdNrp-WfSD-U1M0muEKKaCKr)

![alt text](https://drive.google.com/uc?id=1t4OeIhYuoTzWz8bWvQcYeZUBpqfD6qZG)

![alt text](https://drive.google.com/uc?id=1wzQ50uMZGClHwx22jh3JhQ7IPKHBQ1Fj)

![alt text](https://drive.google.com/uc?id=1akIPl0Sszuy-VYA9OeTjFmqKNIfUJUBJ)

## Coding Time

Python's built-in class provides support for complex numbers.

Complex numbers are defined in rectangular coordinates (real and imaginary parts) using either the constructor or a literal expression.

The complex number 1 + 2j can be defined in either of these ways:

In [0]:
a = complex(1, 2)
b = 1 + 2j
a == b

True

Note that the real and imaginary parts are defined as floats, and can be retrieved as follows:

In [0]:
a.real, type(a.real)

(1.0, float)

In [0]:
a.imag, type(a.imag)

(2.0, float)

The complex conjugate can be calculated as follows:

In [0]:
a.conjugate()

(1-2j)

The standard arithmetic operatots are polymorphic and defined for complex numbers

In [0]:
a = 1 + 2j
b = 3 - 4j
c = 5j
d = 10
a + b

(4-2j)

In [0]:
b * c

(20+15j)

In [0]:
c / d

0.5j

In [0]:
d - a

(9-2j)

The // and % operators, although also polymorphic, are not defined for complex numbers:

In [0]:
a // b

TypeError: ignored

In [0]:
a % b

TypeError: ignored

The == and != operators support complex numbers - but since the real and imaginary parts of complex numbers are floats, the same problems comparing floats using == and != also apply to complex numbers.

In [0]:
a = 0.1j
a + a + a == 0.3j

False

In [0]:
format(a.imag, '.25f')

'1.0000000000000000000000000'

In [0]:
format((a+a+a).imag, '.25f')

'3.0000000000000000000000000'

In [0]:
format((0.3j).imag,'.25f')

'0.2999999999999999888977698'

In addition, the standard comparison operators (<, <=, >, >=) are not defined for complex numbers.

In [0]:
a = 1 + 1j
b = 100 + 100j
a < b

TypeError: ignored

### Math Functions

The **cmath** module provides complex alternatives to the standard **math** functions.
</br>
</br>
In addition, the **cmath** module provides the complex implementation of the **isclose()** method available for floats.

In [0]:
import cmath

a = 1 + 5j
print(cmath.sqrt(a))

(1.7462845577958914+1.4316108957382214j)


The standard **math** module functions will not work with complex numbers:

In [0]:
import math
print(math.sqrt(a))

TypeError: ignored

### Polar / Rectangular Conversions

The **cmath.phase()** function can be used to return the phase (or argument) of  any complex number.
</br>
</br>
The standard **abs()** function supports complex numbers and will return the magnitude (euclidean norm) of the complex number.

In [0]:
a = 1 + 1j
r = abs(a)
phi = cmath.phase(a)
print('{0} = ({1},{2})'.format(a, r, phi))

(1+1j) = (1.4142135623730951,0.7853981633974483)


Complex numbers in polar coordinates can be converted to rectangular coordinates using the **math.rect()** function:

In [0]:
r = math.sqrt(2)
phi = cmath.pi/4
print(cmath.rect(r, phi))

(1.0000000000000002+1j)


### Euler's Identity and the **isclose()** function

e<sup>i &pi;</sup> + 1 = 0

In [0]:
RHS = cmath.exp(cmath.pi * 1j) + 1
print(RHS)

1.2246467991473532e-16j


Which, because of limited precision is not quite zero.
</br>
</br>
However, the result is very close to zero.

We can use the **isclose()** method of the **cmath** module, which behaves similarly to the **math.isclose()** method. Since we are testing for closeness of two numbers close to zero, we need to make sure an absolute tolerance is also specified:

In [0]:
cmath.isclose(RHS, 0, abs_tol=0.00001)

True

If we had not specified an absolute tolerance:

In [0]:
cmath.isclose(RHS, 0)

False

# 15.BOOLEANS

![alt text](https://drive.google.com/uc?id=1f-TSoF4xKQDLsfX5Ouh_NZ58libDeUVR)

![alt text](https://drive.google.com/uc?id=1kqNyW1TYc-EMw2fIdtt-TvP1DK3APm7x)

![alt text](https://drive.google.com/uc?id=1XDW4B5Z3RKsgTSKQcnthlfrIDMfhbltY)

![alt text](https://drive.google.com/uc?id=1tHXQBZehji56ndqSjpCyAZbCjgjZc4he)

![alt text](https://drive.google.com/uc?id=1uFG7YlgIqgLVhx82oKCyL1NNUjLrGmRD)

## Coding Time

The **bool** class is used to represent boolean values.
</br>
</br>
The **bool** class inherits from the **int** class.

In [1]:
issubclass(bool, int)

True

Two built-in constants, **True** and **False** are singleton instances of the bool class with underlying int values of 1 and 0 respectively.

In [2]:
type(True), id(True), int(True)

(bool, 10302848, 1)

In [3]:
type(False), id(False), int(False)

(bool, 10300256, 0)

These two values are instances of the **bool** class, and by inheritance are also **int** objects.

In [4]:
isinstance(True, bool)

True

In [5]:
isinstance(True, int)

True

Since **True** and **False** are singletons, we can use either the **is** operator, or the **==** operator to compare them to **any** boolean expression.

In [6]:
id(True), id(1 < 2)

(10302848, 10302848)

In [7]:
id(False), id(1 == 3)

(10300256, 10300256)

In [8]:
(1 < 2) is True, (1 < 2) == True

(True, True)

In [9]:
(1 == 2) is False, (1 == 2) == False

(True, True)

Be careful with that last comparison, the parentheses are necessary!

In [10]:
1 == 2 == False

False

In [11]:
(1 == 2) == False

True

We'll look into this in detail later, but, for now, this happens because a chained comparison such as **a == b == c** is actually evaluated as **a == b and b == c**
</br>
</br>
So **1 == 2 == False ** is the same as **1 == 2 and 2 == False**

In [12]:
1 == 2, 2 == False, 1==2 and 2==False

(False, False, False)

In [13]:
#BUT
(1 == 2)

False

So **(1 == 2) == False** evaluates to True
</br>
</br>
But since **False** is also **0**, we get the following:

In [14]:
(1 == 2) == 0

True

The underlying integer values of True and False are:

In [15]:
int(True), int(False)

(1, 0)

So, using an equality comparison:

In [16]:
1 == True, 0 == False

(True, True)

But, from an object perspective 1 and True are not the same (similarly with 0 and False)

In [17]:
1 == True, 1 is True

(True, False)

In [18]:
0 == False, 0 is False

(True, False)

Any integer can be cast to a boolean, and follows the rule:

bool(x) = True for any x except for zero which returns False

In [19]:
bool(0)

False

In [20]:
bool(1), bool(100), bool(-1)

(True, True, True)

Since booleans are subclassed from integers, they can behave like integers, and because of polymorphism all the standard integer operators, properties and methods apply

In [21]:
True > False

True

In [22]:
True + 2

3

In [23]:
False // 2

0

In [24]:
True + True + True

3

In [25]:
(True + True + True) % 2

1

In [26]:
-True

-1

In [27]:
100 * False

0

I certainly **do not** recommend you write code like that shown above, but be aware that it does work.

# 16.BOOLEANS: Truth values

![alt text](https://drive.google.com/uc?id=1jtrurT3hpWIILo4yu-QblNfZOj8VkdFO)

![alt text](https://drive.google.com/uc?id=11xfHg7xFoJukCnEk8w41o1IMcb96anSc)

![alt text](https://drive.google.com/uc?id=1A26qtpgTkr-dGfRKs-Kxl8iCwzB4FsSY)

## Coding Time

All objects in Python have an associated **truth value**, or **truthyness**
</br>
</br>
We saw in a previous lecture that integers have an inherent truth value:

In [28]:
bool(0)

False

In [29]:
bool(1), bool(-1), bool(100)

(True, True, True)

This truthyness has nothing to do with the fact that **bool** is a subclass of **int**.

Instead, it has to do with the fact that the **int** class implements a `__bool__()` method

In [30]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __str__(self, /)
 |      Return str(self).
 |  
 |  __xor__(self, value, /)
 |      Return self^val

If you scroll down in the documentation you shoudl reach a section that looks like this:
</br>
</br>
`` 
|  __bool__(self, /)
|      self != 0
``
</br>
</br>
So, when we write:

In [31]:
bool(100)

True

Python is actually calling 100.__bool__() and returning that:

In [32]:
(100).__bool__()

True

In [33]:
(0).__bool__()

False

Most objects will implement either the `__bool__()` or `__len__()` methods. If they don't, then their associated value will be **True** always.

### Numeric Types

Any non-zero numeric value is truthy. Any zero numeric value is falsy:

In [34]:
from fractions import Fraction
from decimal import Decimal
bool(10), bool(1.5), bool(Fraction(3, 4)), bool(Decimal('10.5'))

(True, True, True, True)

In [35]:
bool(0), bool(0.0), bool(Fraction(0,1)), bool(Decimal('0')), bool(0j)

(False, False, False, False, False)

### Sequence Types

An empty sequence type object is Falsy, a non-empty one is truthy:

In [36]:
bool([1, 2, 3]), bool((1, 2, 3)), bool('abc'), bool(1j)

(True, True, True, True)

In [37]:
bool([]), bool(()), bool('')

(False, False, False)

### Mapping Types

Similarly, an empty mapping type will be falsy, a non-empty one truthy:

In [38]:
bool({'a': 1}), bool({1, 2, 3})

(True, True)

In [39]:
bool({}), bool(set())

(False, False)

### The None Object

The singleton **None** object is always falsy:

In [40]:
bool(None)

False

### One Application of Truth Values

Any conditional expression which involves objects other than **bool** types, will use the associated truth value as the result of the conditional expression.

In [41]:
a = [1, 2, 3]
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

1


In [42]:
a = []
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

a is None, or a is empty


In [43]:
a = 'abc'
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

a


In [44]:
a = ''
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

a is None, or a is empty


We could write this using a more lengthy expression:

In [45]:
a = 'abc'
if a is not None and len(a) > 0:
    print(a[0])
else:
    print('a is None, or a is empty')

a


Doing the following would break our code in some instances:

In [46]:
a = 'abc'
if a is not None:
    print(a[0])

a


works, but:

In [47]:
a = ''
if a is not None:
    print(a[0])

IndexError: ignored

In [48]:
a = None
if len(a) > 0:
    print(a[0])

TypeError: ignored

To be torough we would need to write:

In [0]:
a = None
if a is not None and len(a) > 0:
    print(a[0])

Also, the order of the boolean expressions matter here!

We'll discuss this and short-circuit evaluations in an upcoming video.

For example:

In [50]:
a = None
if len(a) > 0 and a is not None:
    print(a[0])

TypeError: ignored